File: internal/mcpauth/hmac.go
package mcpauth

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base64"
	"encoding/hex"
	"errors"
	"net/http"
	"strconv"
	"strings"
	"sync"
	"time"
)

// Request authentication between this server and an attested reader.
//
// The hosted reader shares this host and a bearer secret; an attested reader
// is on another machine, behind a parent instance that sees every byte of the
// traffic TLS does not hide and every header nginx logs. So every request in
// either direction is signed over its method, target, body and a fresh nonce,
// with a direction so that a request signed for one side cannot be replayed
// to the other, and a receiver refuses anything stale or already seen. A
// bearer secret would be replayable by anyone who ever saw one request.

// Directions, as they appear in the canonical string.
const (
	DirectionToReader = "to-reader"
	DirectionToGo     = "to-go"
)

// The headers every signed request carries.
const (
	HeaderReader    = "X-Wappie-Reader"
	HeaderTimestamp = "X-Wappie-Timestamp"
	HeaderNonce     = "X-Wappie-Nonce"
	HeaderSignature = "X-Wappie-Signature"
)

const (
	hmacScheme = "wappie-mcp-hmac/v1"
	// hmacSkew is how far a timestamp may be from this server's clock.
	hmacSkew = 60 * time.Second
	// replayLifetime is how long a nonce is remembered past its timestamp:
	// one second longer than any timestamp stays acceptable.
	replayLifetime = hmacSkew + time.Second
	// replayCap bounds the nonces held at once, after expired ones are
	// swept. A reader at its rate limits sends a few per second; a hundred
	// thousand live nonces is a flood, and the answer to a flood is 503.
	replayCap = 100_000
	// nonceLen is 16 random bytes in unpadded base64url.
	nonceLen = 22
)

// Signature computes the v1 signature of one request: HMAC-SHA256, keyed with
// the secret's bytes as written, over the canonical string. The target is
// the raw path and query exactly as on the request line.
func Signature(secret, direction, readerID, method, target, timestamp, nonce string, body []byte) string {
	sum := sha256.Sum256(body)
	canonical := strings.Join([]string{
		hmacScheme, direction, readerID, strings.ToUpper(method), target, timestamp, nonce, hex.EncodeToString(sum[:]),
	}, "\n")
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(canonical))
	return "v1=" + hex.EncodeToString(mac.Sum(nil))
}

// sign adds the four headers to an outgoing request. The target signed is
// the one Go writes on the request line, so the receiver sees the same bytes.
func sign(req *http.Request, secret, readerID string, body []byte, now time.Time) error {
	nonce, err := newNonce()
	if err != nil {
		return err
	}
	timestamp := strconv.FormatInt(now.Unix(), 10)
	req.Header.Set(HeaderReader, readerID)
	req.Header.Set(HeaderTimestamp, timestamp)
	req.Header.Set(HeaderNonce, nonce)
	req.Header.Set(HeaderSignature, Signature(secret, DirectionToReader, readerID, req.Method, req.URL.RequestURI(), timestamp, nonce, body))
	return nil
}

func newNonce() (string, error) {
	raw := make([]byte, 16)
	if _, err := rand.Read(raw); err != nil {
		return "", err
	}
	return base64.RawURLEncoding.EncodeToString(raw), nil
}

// signedHeaders is what a received request claims about itself.
type signedHeaders struct {
	timestamp string
	unix      int64
	nonce     string
	signature string
}

// errHMACMissing and errHMACStale are the two ways the headers alone fail.
var (
	errHMACMissing = errors.New("hmac_missing")
	errHMACStale   = errors.New("hmac_stale")
)

// readSignedHeaders checks that the timestamp, nonce and signature are each
// present once and well formed, and that the timestamp is within the skew.
// The reader header is the caller's to check, before this.
func readSignedHeaders(h http.Header, now time.Time) (signedHeaders, error) {
	one := func(name string) (string, bool) {
		values := h.Values(name)
		return strings.Join(values, ""), len(values) == 1
	}
	timestamp, okT := one(HeaderTimestamp)
	nonce, okN := one(HeaderNonce)
	signature, okS := one(HeaderSignature)
	if !okT || !okN || !okS || !validTimestamp(timestamp) || !validNonce(nonce) || !validSignature(signature) {
		return signedHeaders{}, errHMACMissing
	}
	unix, err := strconv.ParseInt(timestamp, 10, 64)
	if err != nil {
		return signedHeaders{}, errHMACMissing
	}
	if skew := now.Unix() - unix; skew > int64(hmacSkew/time.Second) || skew < -int64(hmacSkew/time.Second) {
		return signedHeaders{}, errHMACStale
	}
	return signedHeaders{timestamp: timestamp, unix: unix, nonce: nonce, signature: signature}, nil
}

// validTimestamp is decimal Unix seconds with no sign and no leading zero.
func validTimestamp(s string) bool {
	if s == "" || len(s) > 18 || s[0] == '0' {
		return false
	}
	return !strings.ContainsFunc(s, func(c rune) bool { return c < '0' || c > '9' })
}

func validNonce(s string) bool {
	return len(s) == nonceLen && !strings.ContainsFunc(s, func(c rune) bool { return !base64urlChar(c) })
}

func validSignature(s string) bool {
	digest, ok := strings.CutPrefix(s, "v1=")
	return ok && len(digest) == 64 && isHex(digest)
}

// signedBy reports whether any of the secrets produced the signature. Every
// secret is tried, with no early exit, so the time taken does not say which
// one matched or how close a guess came.
func signedBy(secrets []string, got signedHeaders, direction, readerID, method, target string, body []byte) bool {
	matched := 0
	for _, secret := range secrets {
		want := Signature(secret, direction, readerID, method, target, got.timestamp, got.nonce, body)
		matched |= subtle.ConstantTimeCompare([]byte(want), []byte(got.signature))
	}
	return matched == 1
}

// replayCache remembers the nonces of accepted requests until they could no
// longer be accepted anyway. It is filled only after a signature checks, so
// nobody without a secret can put anything in it.
type replayCache struct {
	mu       sync.Mutex
	seen     map[string]time.Time
	capacity int
}

// errReplay and errReplayFull are the two refusals.
var (
	errReplay     = errors.New("hmac_replay")
	errReplayFull = errors.New("replay_cache_full")
)

func newReplayCache(capacity int) *replayCache {
	return &replayCache{seen: map[string]time.Time{}, capacity: capacity}
}

// admit records a nonce, or says it was seen or that there is no room. A
// full cache fails closed: refusing a genuine request is a retry, and
// accepting one unremembered would be a replay window.
func (c *replayCache) admit(direction, readerID, nonce string, timestamp int64, now time.Time) error {
	key := direction + "\x00" + readerID + "\x00" + nonce
	c.mu.Lock()
	defer c.mu.Unlock()
	if until, ok := c.seen[key]; ok && until.After(now) {
		return errReplay
	}
	if len(c.seen) >= c.capacity {
		for k, until := range c.seen {
			if !until.After(now) {
				delete(c.seen, k)
			}
		}
		if len(c.seen) >= c.capacity {
			return errReplayFull
		}
	}
	c.seen[key] = time.Unix(timestamp, 0).Add(replayLifetime)
	return nil
}

File: packages/mcp-http/enclave/hmac.mjs
// HMAC request authentication between Go and the enclave reader, both
// directions (docs/mcp-enclave.md §4). TLS authenticates responses; these
// headers authenticate requests. The canonical string names the direction, so
// a request signed for one side cannot be replayed to the other, and the raw
// request-target, so nothing re-serialized ever reaches the comparison.
import { createHash, createHmac, randomBytes, timingSafeEqual } from 'node:crypto'

export const HMAC_LABEL = 'wappie-mcp-hmac/v1'
export const MAX_SKEW_S = 60
export const REPLAY_TTL_S = 61
export const REPLAY_MAX = 100_000
export const EMPTY_SHA256 = createHash('sha256').update(Buffer.alloc(0)).digest('hex')
const timestampShape = /^(?:0|[1-9][0-9]{0,15})$/
const nonceShape = /^[A-Za-z0-9_-]{22}$/
const signatureShape = /^v1=[0-9a-f]{64}$/

/** The eight fields joined by \n, no trailing newline. */
export function canonicalString({ direction, readerId, method, target, timestamp, nonce, bodySha256 }) {
  return [HMAC_LABEL, direction, readerId, method.toUpperCase(), target, timestamp, nonce, bodySha256].join('\n')
}

/** `v1=` + lowercase hex HMAC-SHA256 keyed by the UTF-8 bytes of the secret as written. */
export function signature(secret, fields) {
  return 'v1=' + createHmac('sha256', Buffer.from(secret, 'utf8')).update(canonicalString(fields), 'utf8').digest('hex')
}

/** The four headers for an outbound request. `body` is the exact bytes sent. */
export function signedHeaders({ secret, direction, readerId, method, target, body = Buffer.alloc(0), now = Date.now }) {
  const timestamp = String(Math.floor(now() / 1000))
  const nonce = randomBytes(16).toString('base64url')
  const bodySha256 = createHash('sha256').update(body).digest('hex')
  return {
    'x-wappie-reader': readerId, 'x-wappie-timestamp': timestamp, 'x-wappie-nonce': nonce,
    'x-wappie-signature': signature(secret, { direction, readerId, method, target, timestamp, nonce, bodySha256 }),
  }
}

/**
 * Nonces seen with a valid signature, each kept until its timestamp + 61 s.
 * Full after sweeping means refusing (503): an attacker who could fill it must
 * already hold a secret, and forgetting a live nonce would admit a replay.
 */
export function createReplayCache({ max = REPLAY_MAX, now = Date.now } = {}) {
  const seen = new Map()
  function sweep() { const at = now() / 1000; for (const [key, until] of seen) if (until <= at) seen.delete(key) }
  return {
    /** 'fresh' (now remembered), 'replay' or 'full'. */
    admit(key, timestamp) {
      const at = now() / 1000
      const until = seen.get(key)
      if (until !== undefined && until > at) return 'replay'
      if (seen.size >= max) { sweep(); if (seen.size >= max) return 'full' }
      seen.set(key, timestamp + REPLAY_TTL_S)
      return 'fresh'
    },
    size: () => seen.size,
  }
}

const refusal = (status, code, logCode) => Response.json({ code }, { status, headers: { 'Cache-Control': 'no-store', 'x-wappie-code': logCode } })

/**
 * The inbound guard (`internalAuth` for startReader), checking `to-reader`
 * requests. `secrets` is the enclave's holder (secrets.mjs): every accepted
 * secret is tried with no early exit, and a request signed with the current
 * one retires the previous. The request body is already bounded by the
 * router's limit; `info.target` is the raw request-target (Node's req.url).
 * Refusals carry `x-wappie-code` for the log line (internal.mjs strips it).
 */
export function createHmacGuard({ readerId, secrets, now = Date.now, replay = createReplayCache({ now }) }) {
  return async function internalAuth(request, info) {
    const header = name => request.headers.get(name)
    const timestamp = header('x-wappie-timestamp') ?? '', nonce = header('x-wappie-nonce') ?? '', given = header('x-wappie-signature') ?? ''
    if (header('x-wappie-reader') !== readerId || !timestampShape.test(timestamp) || !nonceShape.test(nonce) || !signatureShape.test(given)) return refusal(401, 'unauthorized', 'hmac_missing')
    const seconds = Number(timestamp)
    if (Math.abs(now() / 1000 - seconds) > MAX_SKEW_S) return refusal(401, 'unauthorized', 'hmac_stale')
    if (typeof info?.target !== 'string' || !info.target.startsWith('/')) return refusal(401, 'unauthorized', 'hmac_missing')
    const body = Buffer.from(await request.clone().arrayBuffer())
    const fields = { direction: 'to-reader', readerId, method: request.method, target: info.target, timestamp, nonce, bodySha256: createHash('sha256').update(body).digest('hex') }
    const presented = Buffer.from(given)
    let matched = null
    for (const candidate of secrets.accepted()) {
      const expected = Buffer.from(signature(candidate, fields))
      // Every secret is compared, whatever the earlier ones said.
      if (timingSafeEqual(expected, presented) && matched === null) matched = candidate
    }
    if (matched === null) return refusal(401, 'unauthorized', 'hmac_bad')
    const admitted = replay.admit(`to-reader\n${readerId}\n${nonce}`, seconds)
    if (admitted === 'full') return refusal(503, 'replay_cache_full', 'replay_cache_full')
    if (admitted === 'replay') return refusal(401, 'unauthorized', 'hmac_replay')
    secrets.used(matched)
    return undefined
  }
}
